libs/community/langchain_community/document_loaders/sqldatabase.py

```diff
@@ -0,0 +1,135 @@
+from typing import Any, Callable, Dict, Iterator, List, Optional, Union
+
+import sqlalchemy as sa
+
+from langchain_community.docstore.document import Document
+from langchain_community.document_loaders.base import BaseLoader
+from langchain_community.utilities import SQLDatabase
+
+
+class SQLDatabaseLoader(BaseLoader):
+    """
+    Load documents by querying database tables supported by SQLAlchemy.
+
+    For talking to the database, the document loader uses the `SQLDatabase`
+    utility from the LangChain integration toolkit.
+
+    Each document represents one row of the result.
+    """
+
+    def __init__(
+        self,
+        query: Union[str, sa.Select],
```

Comment:
I'm reviewing this PR: https://github.com/langchain-ai/langchain/pull/16655 , which implements a cursor and I think will be used by this data loader for loading documents from SQL.

I've mostly used keyset pagination in the past because it's easier to add retry logic to it in case of network connections or some other server side errors (as well as parallelization).  Is there  way to do retries with the cursor if the connection times out or we get disconnected for whatever reason?

Key set pagination: https://www.cockroachlabs.com/docs/stable/pagination#keyset-pagination

Basic idea is to combine a sort on an id field, a filter on the id filed and a limit.

Reply:
Hi @eyurtsev. I hear your argument, but at the same time, I think the Cockroach documentation is misleading.

Allocating resources for a cursor once, and then using it to churn through the results, optionally also in async/streaming mode, is far more efficient than sending multiple independent database queries which need to allocate corresponding resources on the server side each and every time.

Looking at this from a different perspective, it probably depends very much on the use case at hand. It think using key set pagination is out of scope for this patch, as it depends on an id field, and some database tables may not have it.

If such functionality is needed, it can probably implemented on top, or orthogonal to this infrastructure code, which is merely foundational.

Reply:
>> I've mostly used keyset pagination in the past.
>
> If such functionality is needed, it can probably implemented on top, or orthogonal to this infrastructure code, which is merely foundational.

Hi @eyurtsev,

let me know if you think differently, or if you agree with this assessment. If you are using this technique, and would like to see any sort of adapter for it, I think it may be valuable to implement. However, I don't see how it can be implemented within the scope of this patch.

With kind regards,
Andreas.

Reply:
I'll try to review / merge today if not then earliest on Monday -- a bit pressed on time and I want to review carefully :)

I agree with your points -- cursor based pagination is likely good fit for most users and will result in a simpler API.

If they're working at scale, they'll likely need to do additional optimizations regardless and may want to implement their own solution regardless.

----

fwiw keyset pagination is a general pagination technique (isn't a cockroachdb thing) -- depending on infra and use case it can be used to parallelize reads easily for larger scales of records (~say 10M records), and it's fault tolerant because subsets of pages can be retried independently

Reply:
Thanks Eugene, please take your time. Also thanks for educating me about keyset pagination, I like the idea and it is definitively worth to consider when needing to serve data at scale.

Probably a bit OT, so we might want to continue discussing at another spot, but you sparked my interest ;]:
Is keyset pagination also used when needing to page through datasets which are larger-than-memory, so that processing a potential large resultset in response to a query will not trip memory exhaustion both server- and client-side?

Reply:
> [...] so that processing a potential large resultset in response to a query will not trip memory exhaustion both server- and client-side.

On this very matter, I think it will be sweet to implement server-side cursor support, when applicable. PostgreSQL can do it, and CrateDB implemented it recently, also for that very purpose. See [async_streaming.py](https://github.com/crate/cratedb-examples/blob/main/by-language/python-sqlalchemy/async_streaming.py).

Reply:
Apologies took a bit longer than planned. PR Looks great!  I left a few nits!

Also wanted to thank @cbornet for his review

Reply:
Hi Eugene. Thanks for your review. I will be afk for hiking starting tomorrow, so I may be able to get back to this only on Monday ff. If you have some spare cycles, feel free to take over any time. I don't have any objections about your suggestions. Otherwise, see you next week.